Modules/_threadmodule.c

```diff
@@ -73,13 +74,15 @@ acquire_timed(PyThread_type_lock lock, _PyTime_t timeout)
             /* If we're using a timeout, recompute the timeout after processing
              * signals, since those can take time.  */
             if (timeout > 0) {
-                timeout = endtime - _PyTime_GetMonotonicClock();
+                _PyTime_t elapsed = _PyTime_GetMonotonicClock() - starttime;
 
-                /* Check for negative values, since those mean block forever.
-                 */
-                if (timeout < 0) {
+                if (timeout < elapsed) {
                     r = PY_LOCK_FAILURE;
                 }
+                else {
+                    timeout -= elapsed;
+                    starttime += elapsed;
```

Comment:
I'm not sure that this part is really needed. I would prefer to leave starttime and timeout unchanged.

Reply:
Thank you, @vstinner, for your review.

Reply:
The original code also updates _timeout_ every time it is interrupted. If you don’t adjust the timeout, I think the overall timeout could be too long. I expect you would be able to test this by installing a Python signal handler and sending signals while _acquire_ or whatever method is running.

Perhaps you would prefer to calculate (overall) _timeout_ − (overall) _elapsed_ and pass that to _PyThread_acquire_lock_timed_.

Reply:
`starttime += elapsed` will cause `starttime` overflow. Because `starttime` is closer to `_PyTime_GetMonotonicClock() + timeout` by accumulating, and `_PyTime_GetMonotonicClock() + timeout` may overflow.

I think the `else` block is not needed.